WordPress/gutenberg
# DropdownMenu: refactor to TypeScript

coauthored with @ciampo, who did most of the heavy lifting 💪 

## What?
Refactor `DropdownMenu` component to TypeScript

Part of https://github.com/WordPress/gutenberg/issues/35744

## Why?

The refactor to TypeScript has many benefits (auto-generated docs, static linting and error prevention, better IDE experience). See https://github.com/WordPress/gutenberg/issues/35744 for more details

## How?
Followed the steps in the [TypeScript migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2022-11-29T14%3A45%3A34Z#refactoring-a-component-to-typescript)

## Notes

One decision we had to make while typing the component's `toggleProps` prop was which set of props to import from `Button`. We opted for `ButtonAsButtonProps` because semantically we felt like a `button` made more sense than an `anchor`. I looked at current implementations of `DropDownMenu` and didn't see any passing an `href` to the toggle so should be fine with existing usages. 

In my accessibility searches it didn't look like using a `button` for dropdown menus was a problem but I'd love a more experienced opinion... @andrewhayward do you have any thoughts? 😄 

## Testing Instructions
- Confirm all tests still pass
- In Storybook confirm the component and docs work as expected
